blockindex/liquidstaking_indexer.go

```diff
@@ -0,0 +1,958 @@
+	dirty.putHeight(blk.Height())
+
+	// make action map
+	actionMap := make(map[hash.Hash256]*action.SealedEnvelope)
+	for i := range blk.Actions {
+		h, err := blk.Actions[i].Hash()
+		if err != nil {
+			return err
+		}
+		actionMap[h] = &blk.Actions[i]
+	}
+
+	// handle events of block
+	for _, receipt := range blk.Receipts {
+		if receipt.Status != uint64(iotextypes.ReceiptStatus_Success) {
+			continue
+		}
+		act, ok := actionMap[receipt.ActionHash]
+		if !ok {
+			return errors.Errorf("action %x not found", receipt.ActionHash)
+		}
+		for _, log := range receipt.Logs() {
+			if log.Address != LiquidStakingContractAddress {
+				continue
+			}
+			if err := s.handleEvent(ctx, dirty, blk, act, log); err != nil {
+				return err
+			}
+		}
+	}
+
+	// commit dirty cache
+	return s.commit(dirty)
+}
+
+// DeleteTipBlock deletes the tip block from indexer
+func (s *liquidStakingIndexer) DeleteTipBlock(context.Context, *block.Block) error {
+	return errors.New("not implemented")
+}
+
+// Height returns the tip block height
+func (s *liquidStakingIndexer) Height() (uint64, error) {
+	return s.cache.getHeight(), nil
+}
+
+// CandidateVotes returns the candidate votes
+func (s *liquidStakingIndexer) CandidateVotes(candidate address.Address) *big.Int {
+	return s.cache.getCandidateVotes(candidate)
+}
+
+// Buckets returns the buckets
+func (s *liquidStakingIndexer) Buckets() ([]*Bucket, error) {
+	vbs := []*Bucket{}
+	for id, bi := range s.cache.getAllBucketInfo() {
+		bt := s.cache.mustGetBucketType(bi.TypeIndex)
+		vb, err := s.convertToVoteBucket(id, bi, bt)
+		if err != nil {
+			return nil, err
+		}
+		vbs = append(vbs, vb)
+	}
+	return vbs, nil
+}
+
+// Bucket returns the bucket
+func (s *liquidStakingIndexer) Bucket(id uint64) (*Bucket, error) {
+	bi, ok := s.cache.getBucketInfo(id)
+	if !ok {
+		return nil, errors.Wrapf(ErrBucketInfoNotExist, "id %d", id)
+	}
+	bt := s.cache.mustGetBucketType(bi.TypeIndex)
+	vb, err := s.convertToVoteBucket(id, bi, bt)
+	if err != nil {
+		return nil, err
+	}
+	return vb, nil
+}
+
+// BucketsByIndices returns the buckets by indices
+func (s *liquidStakingIndexer) BucketsByIndices(indices []uint64) ([]*Bucket, error) {
+	vbs := make([]*Bucket, 0, len(indices))
+	for _, id := range indices {
+		bi, ok := s.cache.getBucketInfo(id)
+		if !ok {
+			return nil, errors.Wrapf(ErrBucketInfoNotExist, "id %d", id)
+		}
+		bt := s.cache.mustGetBucketType(bi.TypeIndex)
+		vb, err := s.convertToVoteBucket(id, bi, bt)
+		if err != nil {
+			return nil, err
+		}
+		vbs = append(vbs, vb)
+	}
+	return vbs, nil
+}
+
+func (s *liquidStakingIndexer) TotalBucketCount() uint64 {
+	return s.cache.getTotalBucketCount()
+}
+
+func (s *liquidStakingIndexer) ActiveBucketTypes() (map[uint64]*BucketType, error) {
+	return s.cache.getActiveBucketType(), nil
+}
+
+func (s *liquidStakingIndexer) handleEvent(ctx context.Context, dirty *liquidStakingDirty, blk *block.Block, act *action.SealedEnvelope, log *action.Log) error {
```

Comment:
the purpose of act as one of the arguments?

Reply:
`act` has been an unused argument and should be removed